promql/functions.go

```diff
@@ -810,11 +810,14 @@ func funcHistogramQuantile(vals []parser.Value, args parser.Expressions, enh *Ev
 			continue
 		}
 		l := sigf(el.Metric)
+		// Add the metric name (which is always removed) to the signature to prevent combining multiple histograms
+		// with the same label set. See https://github.com/prometheus/prometheus/issues/9910
+		l = l + el.Metric.Get(model.MetricNameLabel)
```

Comment:
Beyond hackiness, a minor concern here is that this creates a whole new string each time. If that's a relevant performance drain, we would need to add the name label before calling the `String` method of the label set.